libvast/vast/community_id.hpp

```diff
@@ -0,0 +1,207 @@
+/// @param resp_h The IP address of the flow destination.
+/// @param orig_p The transport-layer port of the flow source.
+/// @param resp_p The transport-layer port of the flow destination.
+/// @return An instance of a flow.
+/// @relates flow
+template <port::port_type Protocol>
+flow make_flow(address orig_h, uint16_t orig_p, address resp_h,
+               uint16_t resp_p) {
+  return flow{orig_h, resp_h, port{orig_p, Protocol}, port{resp_p, Protocol}};
+}
+
+/// Factory function to construct a flow.
+/// @param orig_h The IP address of the flow source.
+/// @param resp_h The IP address of the flow destination.
+/// @param orig_p The transport-layer port of the flow source.
+/// @param resp_p The transport-layer port of the flow destination.
+/// @relates flow
+template <port::port_type Protocol>
+caf::optional<flow> make_flow(std::string_view orig_h, uint16_t orig_p,
+                              std::string_view resp_h, uint16_t resp_p) {
+  using parsers::addr;
+  flow result;
+  if (!addr(orig_h, result.src_addr) || !addr(resp_h, result.dst_addr))
+    return caf::none;
+  result.src_port = port{orig_p, Protocol};
+  result.dst_port = port{resp_p, Protocol};
+  return result;
+}
+
+/// Computes the length of the version prefix.
+/// @see max_length
+inline constexpr size_t version_prefix_length() {
+  constexpr size_t version_number = 1;
+  constexpr size_t version_separator = 1;
+  return version_number + version_separator;
+}
+
+/// Computes the maximum length of the Community ID string.
+/// @returns An upper bound in bytes.
+/// @see version_prefix_length
+template <class Policy>
+constexpr size_t max_length() {
+  constexpr auto hex_size = (160 / 8) * 2; // 160-bit SHA-1 digest as hex.
+  auto prefix = version_prefix_length();
+  if constexpr (std::is_same_v<Policy, policy::base64>)
+    return prefix + detail::base64::encoded_size(hex_size);
+  else if constexpr (std::is_same_v<Policy, policy::ascii>)
+    return prefix + hex_size;
+  return 0;
+}
+
+/// Calculates the Community ID for a given flow.
+/// @tparam Policy The rendering policy to select Base64 or ASCII.
+/// @param x The flow tuple.
+/// @param seed An optional seed to the SHA-1 hash.
+/// @returns A string representation of the Community ID for *x*.
+template <class Policy>
+std::string compute(const flow& x, uint16_t seed = 0) {
+  std::string result;
+  // Perform exactly one allocator round-trip.
+  result.reserve(max_length<Policy>());
+  // The version prefix is always present.
+  result += version;
+  result += ':';
+  // Compute a SHA-1 hash over the flow tuple.
+  sha1 hasher;
+  hash_append(hasher, detail::to_network_order(seed));
+  hash_append(hasher, x);
+  auto digest = static_cast<sha1::result_type>(hasher);
+  // Convert the binary digest to plain hex ASCII or to Base64.
+  constexpr auto element_size = sizeof(sha1::result_type::value_type);
+  constexpr auto num_bytes = element_size * digest.size();
```

Comment:
I'm surprised this compiles. `digest` is not a `constexpr` value, so logically `num_bytes` can't be either.

```suggestion
  auto num_bytes = element_size * digest.size();
```

Reply:
Actually, [`array<T, N>::size`](https://en.cppreference.com/w/cpp/container/array/size) *is* `constexpr`, so this works. This makes also sense intuitively, because the size is known at compile-time.